### DISCUSSION TITLE
Driving an oscillating stepper motor as a rocker for baby swing with Tasmota (UI)

### DISCUSSION THREAD

**Discussion by karlitos** [arendst/Tasmota#16012] [Q&A] [answered]
Hello dear tasmota community,

I became a father of a beautiful baby girl. We keep her in a baby swing ![Baby swing] and she is loving it !  ![Baby swing](https://www.wunderwiege.de/media/catalog/product/cache/a26f3c6bba10627ca3d0ca19c4f69980/s/e/set-1_wolli_single_beige.webp)

I flashed Tasmota to an ESP CAM module and attached it to the spreader bar so I can monitor her in her sleep from another room. The module has plenty of spare pins and I would love to drive a stepper motor which would rock the swing gently up and down.

For this I already bought an [TMC2209](https://wiki.fysetc.com/Silent2209/) silent stepper driver, and was able to test it with some Arduino code - it runs absolutely noiseless! (Which is kind of a must have in my case 😄  )

I found already some articles about driving a stepper motor with Tasmota ([here](https://github.com/arendst/Tasmota/discussions/11731) and [here](https://tasmota.github.io/docs/A4988-Stepper-Motor-Controller/)) but they do not entirely cover my intend.

I need the stepper motor to oscillate - back and forth - meaning it should turn certain amount of steps in one direction and then certain amount of steps in the other direction. There will be a small arm attached to the shaft, which only needs to move few cm up and down (The baby should be gently rocked, not get a county fair attraction ride).

Depending on the length of the arm, the desired amplitude and frequency can be translated to driving a step and direction pins on the motor controller.

__My question is:__

1. How to add an UI - sliders / input fields, which would allow me to set some values as  amplitude and frequency
2. How to add a button which would - depending of the above values - drive some GPIOs (driving the step-pin of the controller and setting the direction pin high/low)
3. Do I need / should I use the [A4988 stepper motor controller](https://tasmota.github.io/docs/A4988-Stepper-Motor-Controller/) ?

In general I would be very happy if someone could point me to the right direction (maybe [scripting](https://tasmota.github.io/docs/Scripting-Language/) ???), there is just so much documentation and it is hard find the right entry point 🙇‍♂️

**Comment by sfromis** [comment, marked as answer]
The [ESP32-CAM](https://templates.blakadder.com/ai-thinker_ESP32-CAM) board (which I assume that you mean) does not have "plenty of spare pins", as most of the low number of broken out pins are also used for the camera, as visible in the template linked. Another factor is that the camera function is already hard work for the ESP32, and we generally recommend to not add to that.

With ESP32, my preference for adding user interface elements and controls would be to use the Berry language. As I've not worked with motors, I'm not going to try to make uninformed suggestions on that part. More generally, when safety (for the baby) becomes a factor, you should be extra careful that no harm can possibly be done if things stops working, or does unexpected stuff, for whatever reason.

**Comment by joba-1** [comment]
I agree with @sfromis : the esp-cam is all but ideal for the task. It can work in principle, but  it will feel like swimming upstream: for quiet ops you need reliable microseconds timing. Not the domain of esps and not for tasmota either - a busy cam makes it only worse. I‘d add an STM32 (any will do) and let it do the stepper job (controlled by tasmota)

**Comment by karlitos** [comment]
Thank you for your suggestions. I did not know about the berry language, I will definitely check it.

I was unaware about the fact, that the camera function is keeping the board (yes it is the ESP32-CAM) so busy. So far, I just tried to take some camera example code for the ESP32-CAM and add some simple code setting the step pin on the TMC2209 high and low and changing the direction of the rotation. This worked and the operation was silent. I just missed having some basic Web-UI allowing me to change the settings (amplitude, frequency).

In general "will work in principle" will be enough for my case since time is playing against me - having any optimal solution when the baby is not using the swing anymore will fail the purpose 😃 but when adding an extra microcontroller will drastically lower the complexity - why not ? How will the Tasmota communicate with it ? Serial connection/I2C ?

Does it need to be an STM32 or will an Arduino micro/trinket or any miniature board do the job ? I'd like to avoid having bunch of wired components hanging around.

**Comment by joba-1** [comment]
Hm, if the TMC2209 already does the smoothing job good enough, then it already acts as that precisely timed microcontroller helping out the jittery ESPs. I only did "hardcore" stepper driving myself, powering the coils directly. 
With an extra microcontroller, any will do. I don't know another µC as bad with precise timing as the ESPs. I'm just most familiar with the STM32s but if you have experience with another one that you maybe already have then go for it. 
For comms it is similar: any of your suggestions can work - mainly depends on your knowledge. Serial is probably easiest: very flexible in what you can communicate, widely supported by all µC, only one (or two) wires. So I would go for that.
And regarding time: you may have plenty! My son recently asked me to mount a swing seat at the ceiling and probably would not mind a stepper motor doing the swinging for him. He is 22 :)

**Reply by stefanbode** [reply]
You can also use the shutter and blinds driver. Here you can also do some fancy acceleration stuff at the end points. To get the swing I would just create two rule that send the shutter in the opposite direction if endpoint is reached.
